ykrt/src/compile/jitc_yk/codegen/x64/lsregalloc.rs

```diff
@@ -1169,8 +1169,8 @@ impl From<Rx> for RegSet<Rx> {
 enum SpillState {
     /// This variable has not yet been spilt, or has been spilt and will not be used again.
     Empty,
-    /// This variable is spilt to the stack with the same semantics as [VarLocation::Direct].
-    Direct(i32),
+    /// This variable is spilt to the stack with the same semantics as [VarLocation::Stack].
```

Comment:
I think we have to be careful here. Stackmaps have two ways of expressing stack storage: `Direct` and `Indirect`.

https://llvm.org/docs/StackMaps.html#stack-map-format:
> Values spilled to the stack are encoded as Indirect locations. The runtime must load those values from a stack address, typically in the form [BP + Offset]. If an alloca value is passed directly to a stack map intrinsic, then LLVM may fold the frame index into the stack map as an optimization to avoid allocating a register or stack slot. These frame indices will be encoded as Direct locations in the form BP + Offset.

Perhaps `Direct` is more precise after all?

Reply:
As things stand right now, it seems `Stack`, `Direct`, and `Indirect` are all different. *Why* exactly that's the case is one of the things I raised in the recent issue. It feels to me like there should only be 2 states, but maybe we really do need 3. I really don't know. What I do know is that this PR currently only handles `VarLocation::Stack` and so `Spillstate::Stack` is the right name (for now...).

Reply:
I think you will need to ask Lukas if this is correct when he's back. We can merge it for now.

I don't know what `Stack` is though. That's not mentioned in the stackmap docs.

The valid locations are:
```
Register | Direct | Indirect | Constant | ConstantIndex
```

Reply:
`Stack` is the "traditional" yk notion of "the SSA value is on the trace function's frame stack at [rbp - frame_off]".

Reply:
Right, and that is expressed in stackmaps as one of `Direct` or `Indirect`, so a third stack stack shouldn't be needed AFAIK.

Reply:
*Maybe* there is a good reasons for 3. `Stack` is "the yk reg alloc spilled the value" and `Direct` and `Indirect` are "LLVM put this on a stack somewhere". So *maybe* it's just the names that suck. I suspect we shouldn't follow LLVM's terminology here, and we should also rename `Stack` to something else. But this is something that's probably best handled by @ptersilie.

Reply:
Right, I see what the issue is. It's correct that LLVM has 2 ways to describe a value on the stack:
1) Indirect: This is a value that was spilled to the stack, and the pointer is also stored in the stack. So to get the value we need to do an extra indirection: `val = *(rbp - 8)`
2) Direct: This is stackmaps realising the live value is itself a pointer to the stack. So instead of spilling the pointer, stackmaps just records the stack offset: `val = rbp - 8`

However, at the moment the interpreter frame and the compiled-trace frame are separate. And that's why, I believe, Laurie needed to add another `Stack` variant, to describe something that lives on the JIT stack.

This was one of the changes I had planned to do next: to execute the trace on the same stack as the parent frame. Then we don't need that extra variant anymore. Maybe this can be done in this PR, I'm not sure. It "should" be as simple as initialising the register allocator with the size of the main interpreter frame. In other words, we tell the register allocator: here's an existing stack, so the first 10 bytes are already used, anything else you need to push, you push after that. Happy to do that in a follow up PR though.

Reply:
Oh, just saw this PR is already merged. So ignore my last point. :laughing: